Plutarch/TryFrom.hs

```diff
@@ -0,0 +1,367 @@
+        verifyPair = plam $ \tup -> unTermCont $ do
+          (verfst, _) <- tcont $ ptryFrom @(PAsData a) $ pfstBuiltin # tup
+          (versnd, _) <- tcont $ ptryFrom @(PAsData b) $ psndBuiltin # tup
+          pure $ ppairDataBuiltin # verfst # versnd
+    ver <- tcont $ plet $ pmap # verifyPair # verMap
+    pure (punsafeCoerce opq, ver)
+
+{- |
+    This verifies a list to be indeed a list but doesn't recover the inner data
+    use this instance instead of the one for `PData (PAsData (PBuiltinList (PAsData a)))`
+    as this is O(1) instead of O(n)
+-}
+
+-- TODO: add the excess inner type list
+instance PTryFrom PData (PAsData (PBuiltinList PData)) where
+  type PTryFromExcess PData (PAsData (PBuiltinList PData)) = Flip Term (PBuiltinList PData)
+  ptryFrom' opq = runTermCont $ do
+    ver <- tcont $ plet (pasList # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+{- |
+    Recover a `PBuiltinList (PAsData a)`
+-}
+instance
+  ( PTryFrom PData (PAsData a)
+  , PIsData a
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinList (PAsData a)))
+  where
+  type PTryFromExcess PData (PAsData (PBuiltinList (PAsData a))) = Flip Term (PBuiltinList (PAsData a))
+  ptryFrom' opq = runTermCont $ do
+    let lst :: Term _ (PBuiltinList PData)
+        lst = pasList # opq
+        verify :: Term _ (PData :--> PAsData a)
+        verify = plam $ \e ->
+          unTermCont $ do
+            (wrapped, _) <- tcont $ ptryFrom @(PAsData a) $ e
+            pure wrapped
+    ver <- tcont $ plet $ pmap # verify # lst
+    pure $ (punsafeCoerce opq, ver)
+
+{- |
+    Recover a `PAsData (PBuiltinPair a b)`
+-}
+instance
+  ( PTryFrom PData a
+  , a ~ PAsData a'
+  , PIsData a'
+  , PTryFrom PData b
+  , b ~ PAsData b'
+  , PIsData b'
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinPair a b))
+  where
+  type PTryFromExcess PData (PAsData (PBuiltinPair a b)) = Flip Term (PBuiltinPair a b)
+  ptryFrom' opq = runTermCont $ do
+    tup <- tcont $ plet (pfromData $ punsafeCoerce opq)
+    let fst' :: Term _ a
+        fst' = unTermCont $ fst <$> tcont (ptryFrom @a $ pforgetData $ pfstBuiltin # tup)
+        snd' :: Term _ b
+        snd' = unTermCont $ fst <$> tcont (ptryFrom @b $ pforgetData $ psndBuiltin # tup)
+    ver <- tcont $ plet $ ppairDataBuiltin # fst' # snd'
+    pure $ (punsafeCoerce opq, ver)
+
+----------------------- PDataRecord instances -------------------------------------------
+
+-- We could have a more advanced instance but it's not needed really.
+
+newtype ExcessForField (a :: PType) (s :: S) = ExcessForField (Term s (PAsData a), Reduce (PTryFromExcess PData (PAsData a) s))
+
+instance Reducible (PTryFromExcess PData (PAsData a) s) => Reducible (ExcessForField a s) where
+  type Reduce (ExcessForField a s) = (Term s (PAsData a), Reduce (PTryFromExcess PData (PAsData a) s))
+
+-- FIXME: Should we always succede? If we always succede, performance would increase a lot.
```

Comment:
outdated comment?

Reply:
It's not outdated. It's about verifying whether the list is empty or not.

Reply:
well yeah, I mean that there's no way to make this a "always succeeds" no-op right?